### PR TITLE
JDK-8319256: Print more diagnostic information when an unexpected user is found in a Phi

### PR DESCRIPTION
**Pull request by JohnTortugo**
Can I please get some review on this PR to:

- Add more verification around the implementation of reduce allocation merges.
- Print more diagnostic information when an unexpected situation is encountered.

Testing: Linux x64 tier1-4, Renaissance, Dacapo and SpecJBB.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319256](https://bugs.openjdk.org/browse/JDK-8319256): Print more diagnostic information when an unexpected user is found in a Phi (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [70312fb7](https://git.openjdk.org/jdk/pull/16465/files/70312fb7edd91b315e36072794a73bb266470ebd)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [70312fb7](https://git.openjdk.org/jdk/pull/16465/files/70312fb7edd91b315e36072794a73bb266470ebd)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16465/head:pull/16465` \
`$ git checkout pull/16465`

Update a local copy of the PR: \
`$ git checkout pull/16465` \
`$ git pull https://git.openjdk.org/jdk.git pull/16465/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16465`

View PR using the GUI difftool: \
`$ git pr show -t 16465`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16465.diff">https://git.openjdk.org/jdk/pull/16465.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16465#issuecomment-1791643684)

### CODE REVIEW

File: src/hotspot/share/opto/c2_globals.hpp [hunk 1]
```diff
@@ -470,6 +470,9 @@
   notproduct(bool, TraceReduceAllocationMerges, false,                      \
           "Trace decision for simplifying allocation merges.")              \
                                                                             \
+  notproduct(bool, VerifyReduceAllocationMerges , true,                     \
```
Comment (vnkozlov) [inline src/hotspot/share/opto/c2_globals.hpp:473]: From changes I assume you want verification done in debug VM. Use `develop(`  flag declaration for that.
(`notproduct` includes `optimized` VM build).

File: src/hotspot/share/opto/escape.cpp [hunk 2]
```diff
@@ -369,6 +369,18 @@ bool ConnectionGraph::compute_escape() {
       assert(ptn->escape_state() == PointsToNode::NoEscape && ptn->scalar_replaceable(), "sanity");
     }
   }
+
+  if (VerifyReduceAllocationMerges) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:373]: Add `#ifdef ASSERT`

File: src/hotspot/share/opto/escape.cpp [hunk 3]
```diff
@@ -369,6 +369,18 @@ bool ConnectionGraph::compute_escape() {
...
+    for (uint i = 0; i < reducible_merges.size(); i++ ) {
+      Node* n = reducible_merges.at(i);
+      if (!can_reduce_phi(n->as_Phi())) {
+        NOT_PRODUCT(TraceReduceAllocationMerges = true;)
```
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:377]: With `#ifdef` you don't need `NOT_PRODUCT()` here.

File: src/hotspot/share/opto/escape.cpp [hunk 4]
```diff
@@ -530,14 +542,21 @@ void ConnectionGraph::reduce_phi_on_field_access(PhiNode* ophi, GrowableArray<No
   // though the load doesn't have an unique instance type.
   bool ignore_missing_instance_id = true;
 
+#ifdef ASSERT
+  if (VerifyReduceAllocationMerges && !can_reduce_phi(ophi)) {
+    NOT_PRODUCT(TraceReduceAllocationMerges = true;)
```
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:547]: With #ifdef you don't need NOT_PRODUCT() here.

File: src/hotspot/share/opto/escape.cpp [hunk 5]
```diff
@@ -3783,6 +3814,7 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
       if (reducible_merges.member(n)) {
         // Split loads through phi
         reduce_phi_on_field_access(n->as_Phi(), alloc_worklist);
+        if (VerifyReduceAllocationMerges) reduced_merges.push(n);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:3817]: Code style. split line and use `{}` for body. And add `#ifdef ASSERT`.

File: src/hotspot/share/opto/escape.cpp (hunk 2 above)
Comment (JohnTortugo) [reply src/hotspot/share/opto/escape.cpp:373]: This block is already inside an `#ifdef ASSERT`.

**Review by vnkozlov — COMMENTED** [review]
Good. I submitted testing.

**Comment by JohnTortugo** [comment]
Thank you for reviewing & testing @vnkozlov !

File: src/hotspot/share/opto/c2_globals.hpp [hunk 6]
```diff
@@ -470,6 +470,9 @@
   notproduct(bool, TraceReduceAllocationMerges, false,                      \
           "Trace decision for simplifying allocation merges.")              \
                                                                             \
+  develop(bool, VerifyReduceAllocationMerges , true,                        \
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/c2_globals.hpp:473]:
```suggestion
  develop(bool, VerifyReduceAllocationMerges , true,                        \
```
```suggestion
  develop(bool, VerifyReduceAllocationMerges, true,                         \
```

File: src/hotspot/share/opto/c2_globals.hpp [hunk 7]
```diff
@@ -470,6 +470,9 @@
   notproduct(bool, TraceReduceAllocationMerges, false,                      \
           "Trace decision for simplifying allocation merges.")              \
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/c2_globals.hpp:471]:
```suggestion
             "Trace decision for simplifying allocation merges.")           \
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by TobiHartmann** [comment]
@JohnTortugo Due to the whitespace changes, you need to issue the integrate command again.

**Comment by JohnTortugo** [comment]
Thank you for reviewing.

**Pull request closed without merging**